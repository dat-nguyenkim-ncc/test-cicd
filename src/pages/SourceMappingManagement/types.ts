import { FieldTypes } from '../../components/TextField'

export type FundingRoundMappingDTO = {
  sourceValue: string
  round2Id: number
  round1Id: number
  round1: string
  round2: string
  id: number
}

export type FundingRoundMappingResponse = { data: FundingRoundMappingDTO[]; total: number }

export type GetFundingRoundMappingsResponse = {
  getFundingRoundMappings: FundingRoundMappingResponse
}

export type PendingFundingRoundMappingUpdate = Record<
  keyof Omit<FundingRoundMappingDTO, 'id'>,
  string | number
>

export type FundingRoundMappingFieldDTO = {
  field: keyof FundingRoundMappingDTO
  key: keyof FundingRoundMappingDTO
  label: string
  canEdit: boolean
  validate?: (v: string | number) => boolean
  format?: (v: string | number) => string
  fieldType: FieldTypes
  disablePopover?: boolean
  formLabel?: string
}

export type FundingRoundMappingFilterDTO = {
  sourceValue?: string
  round2Id: number
  round1Id: number
  isSourceValueBlank: boolean
}
