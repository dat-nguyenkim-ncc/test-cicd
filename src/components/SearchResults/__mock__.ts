export const data = {
  external: [
    {
      //replace property id by company_id
      //TODO: change all records
      company_id: '7b4545e6-a34d-47c2-853e-ff9eb40c18ef',
      source: 'CRB',
      companyDetails: {
        companyId: '9185a077-c260-ac40-2678-fef9e645591b',
        companyName: 'Monzo',
        url: 'https://monzo.com/',
        countryName: 'United Kingdom',
        countryCode: 'GB',
      },
    },
    {
      id: 'a67f6ebd-c2e4-4c15-8657-4d40cc7626f1',
      source: 'DR',
      companyDetails: {
        companyId: '74473',
        companyName: 'Monzo',
        url: 'https://monzo.com',
        countryName: 'United Kingdom',
        countryCode: 'GB',
      },
    },
    {
      id: '89c20324-1390-4ea2-acbc-7cc05a98db59',
      source: 'CRB',
      companyDetails: {
        companyId: '33164743-a09a-45cb-a160-a54ace4dc3da',
        companyName: 'Monzon Biomed',
        url: 'http://www.monzonbiomed.com/',
        countryName: 'Spain',
        countryCode: 'ES',
      },
    },
    {
      id: '807a6c17-a4ff-4b74-b9ef-fee94364f451',
      source: 'CRB',
      companyDetails: {
        companyId: '456754eb-67c6-98db-780a-327db4c05c8f',
        companyName: 'monzoom.com',
        url: 'https://www.monzoom.com',
        countryName: 'Austria',
        countryCode: 'AT',
      },
    },
    {
      id: '0511b523-c25b-443b-932f-4d501d5c0cd4',
      source: 'CRB',
      companyDetails: {
        companyId: '09c52685-0f60-4a6f-a5fc-5c51ae5bec3a',
        companyName: 'Monzoon Networks',
        url: 'http://hotspot.monzoon.net',
        countryName: 'Switzerland',
        countryCode: 'CH',
      },
    },
  ],
  internal: [
    {
      //replace property id by company_id
      //TODO: change all records
      company_id: '6',
      source: 'MANUAL',
      companyDetails: {
        companyId: '6',
        companyName: 'MonzoBCG',
        url: 'www.MonzoBCG.com',
        countryName: 'England',
        countryCode: null,
      },
    },
    {
      id: '127237',
      source: 'DR',
      companyDetails: {
        companyId: '127237',
        companyName: 'Monzo',
        url: 'https://monzo.com',
        countryName: 'United Kingdom',
        countryCode: 'GB',
      },
    },
    {
      id: '141132',
      source: 'MANUAL',
      companyDetails: {
        companyId: '141132',
        companyName: 'Monzo',
        url: 'https://monzo.com/',
        countryName: 'United Kingdom',
        countryCode: 'GB',
      },
    },
  ],
}
