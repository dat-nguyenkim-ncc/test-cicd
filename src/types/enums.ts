export enum Routes {
  SEARCH = '/search',
  SEARCH_QUERY = '/search/:query',
  ADD_COMPANY = '/add',
  ADD_COMPANY_OVERVIEW = '/add/overview',
  ADD_COMPANY_FINANCIALS = '/add/financials',
  ADD_COMPANY_FUNDRAISING = '/add/fundraising',
  ADD_COMPANY_BUSINESS = '/add/business',
  ADD_COMPANY_PEOPLE = '/add/people',
  ADD_COMPANY_TAXONOMY = '/add/taxonomy',
  ADD_COMPANY_TAXONOMY_EXTERNAL = '/add/:id/taxonomy',
  COMPANY = '/company/:id',
  COMPANY_EDIT_SOURCE = '/company/sources/:id',
  COMPANY_NEW = '/company/new/:id/:source',
  EDIT_COMPANY_TAXONOMY = '/edit/:id/taxonomy',
  EDIT_COMPANY_OVERVIEW = '/edit/:id/overview',
  EDIT_COMPANY_FINANCIALS = '/edit/:id/financials',
  EDIT_COMPANY_INVESTMENTS = '/edit/:id/investments',
  EDIT_COMPANY_FUNDRAISING = '/edit/:id/fundraising',
  EDIT_COMPANY_BUSINESS = '/edit/:id/business',
  EDIT_COMPANY_PEOPLE = '/edit/:id/people',
  EDIT_COMPANY_ACQUISITIONS = '/edit/:id/acquisitions',
  EDIT_COMPANY_ACQUIREES = '/edit/:id/acquirees',
  EDIT_COMPANY_IPOS = '/edit/:id/ipos',
  EDIT_COMPANY_NEWS = '/edit/:id/news',
  EDIT_COMPANY_USE_CASE = '/edit/:id/use-case',
  EDIT_COMPANY_TECHNOLOGY = '/edit/:id/technology',
  EDIT_COMPANY_PARTNERSHIPS = '/edit/:id/partnerships',

  EDIT_COMPANY_FINANCIALS_CR = '/edit/:id/financials/:cr',
  EDIT_COMPANY_FUNDRAISING_CR = '/edit/:id/fundraising/:cr',
  EDIT_COMPANY_ACQUISITIONS_CR = '/edit/:id/acquisitions/:cr',
  EDIT_COMPANY_IPOS_CR = '/edit/:id/ipos/:cr',
  EDIT_COMPANY_NEWS_CR = '/edit/:id/news/:cr',
  EDIT_COMPANY_PEOPLE_CR = '/edit/:id/people/:cr',
  EDIT_COMPANY_TAXONOMY_CR = '/edit/:id/taxonomy/:cr',
  INVESTOR_MANAGEMENT_CR = '/investor-management/:cr',
  EDIT_COMPANY_USE_CASE_CR = '/edit/:id/use-case/:cr',
  EDIT_COMPANY_TECHNOLOGY_CR = '/edit/:id/technology/:cr',
  EDIT_COMPANY_BUSINESS_CR = '/edit/:id/business/:cr',
  EDIT_COMPANY_PARTNERSHIPS_CR = '/edit/:id/partnerships/:cr',

  MERGE = '/merge',
  MERGE_COMPANY = '/merge/:query',
  TAXONOMY_MANAGEMENT = '/taxonomy-management',
  TAG_MANAGEMENT = '/tag-management',
  MAPPING_ZONE = '/mapping-zone',
  REPORT_MANAGEMENT = '/report-management',
  COMPANY_MANAGEMENT = '/company-management',
  BULK_EDIT = '/company-management/bulk-edit',
  HEALTH_CHECK = '/health-check',
  CHANGE_REQUEST_MANAGEMENT = '/change-request-management',
  CHANGE_REQUEST_MANAGEMENT_TAXONOMY = '/taxonomy-change-request',
  INVESTOR_MANAGEMENT = '/investor-management',
  FIND_FINTECHS = '/find-fintechs',
  SIMILAR_COMPANIES = '/similar-companies',
  SOURCE_MAPPING_MANAGEMENT = '/source-mapping-management',
  INCORRECT_MAPPING = '/incorrect-mapping',
  ANALYSIS = '/analysis',
}

const { REACT_APP_STAGE } = process.env
const prefix = REACT_APP_STAGE === 'prod' ? 'prod' : 'dev'

export const EnumUserGroups = {
  KT: `fct_dataplatform_${prefix}_KT`,
  EXPAND: `fct_dataplatform_${prefix}_expand`,
  EVS: `fct_dataplatform_${prefix}_EVS`,
  MISSION_PLUS: `fct_dataplatform_${prefix}_mission_plus`,
}

export enum EnumVariantKeys {
  primary = 'primary',
  secondary = 'secondary',
  outline = 'outline',
  outlineWhite = 'outlineWhite',
  black = 'black',
  muted = 'muted',
  error = 'error',
  invert = 'invert',
}

export enum EnumCompanyTypeSector {
  FIN = 'fintech',
  INS = 'insurtech',
  REG = 'regtech',
  OUT = 'out',
}

export enum EnumDimensionType {
  SECTOR = 'sector',
  CLUSTER = 'cluster',
  VALUE_CHAIN = 'value chain',
  RISH = 'risk',
}

export type DimensionType =
  | EnumDimensionType.SECTOR
  | EnumDimensionType.CLUSTER
  | EnumDimensionType.VALUE_CHAIN
  | EnumDimensionType.RISH

export enum EnumExpandStatus {
  FOLLOWING = 'Following',
  UNFOLLOWED = 'Unfollowed',
  DUPLICATED = 'Duplicated',
  TO_BE_EVALUATED = 'ToBeEvaluated',
  CHANGE_REQUEST = 'ChangeRequest',
}

export enum EnumExpandStatusId {
  FOLLOWING = '1',
  UNFOLLOWED = '2',
  DUPLICATED = '3',
  TO_BE_EVALUATED = '4',
  CHANGE_REQUEST = '5',
}

export enum EnumCompanyNewsStatusId {
  FOLLOWING = '1',
  UNFOLLOWED = '0',
}

export enum EnumCompanyNewsStatus {
  FOLLOWING = 'Following',
  UNFOLLOWED = 'Unfollowed',
}

export enum EnumTagGroupSource {
  BCG = 'bcg',
  DR = 'dealroom',
  CRB = 'crunchbase',
  BCG_FIXED = 'bcg_fixed',
  FEEDLY = 'feedly',
  SWITCHPITCH = 'switchpitch',
}

export enum EnumApiSource {
  CRB = 'CRB',
  DR = 'DR',
  MANUAL = 'MANUAL',
  SWITCHPITCH = 'SWITCHPITCH',
  FEEDLY = 'FEEDLY',
}

export enum EnumReverseApiSource {
  CRB = 'crunchbase',
  DR = 'dealroom',
  MANUAL = 'bcg',
  SWITCHPITCH = 'switchpitch',
  FEEDLY = 'feedly',
}

export type ApiSourceType =
  | EnumApiSource.CRB
  | EnumApiSource.DR
  | EnumApiSource.MANUAL
  | EnumApiSource.SWITCHPITCH
  | EnumApiSource.FEEDLY

export enum RoundStatus {
  FOLLOWING = 'following',
  NOT_FOLLOWING = 'not following',
}

export enum EnumCompanySource {
  BCG = 'bcg',
  DR = 'dealroom',
  CRB = 'crunchbase',
}

export enum EnumReverseCompanySource {
  bcg = 'BCG',
  dealroom = 'DR',
  crunchbase = 'CRB',
}

export enum EnumInvestorSource {
  bcg = 'BCG',
  dealroom = 'DR',
  crunchbase = 'CRB',
  bcg_comp = 'BCG_COMP',
}

export enum EnumRemoveSourceType {
  MAKE_TO_OUT = 1,
  CREATE_NEW_MAPPING = 2,
  KEEP_EXISTING_MAPPING = 3,
  SET_AS_DUPLICATE = 4,
}

export enum EnumMergeStep {
  PRIMARY = 1,
  AUXILIARY = 2,
  FINTECH_TYPE = 3,
  TAGS = 4,
  PROFILE = 5,
  USE_CASE = 6,
  FUNDRAISING = 7,
  TECHNOLOGY = 8,
  LOCATION = 9,
  OVERRIDES = 10,
  SUMMARY = 11,
}

export enum ValidateInput {
  MAX_RANGE_BIGINT = 18446744073709551615,
  MAX_RANGE_INT = 2147483647,
}

export enum EnumDimensionCategories {
  FIN = 'fintech',
  INS = 'insurtech',
  REG = 'regtech',
}

export enum EnumDimensionValue {
  PRIMARY = 1,
  SECONDARY = 2,
}

export enum EnumS3Operation {
  PUT = 'putObject',
  GET = 'getObject',
}
export enum SortDirection {
  ASC = 'ASC',
  DESC = 'DESC',
}

export const PERMISSIONS: Partial<Record<Routes, string[]>> = {
  [Routes.TAG_MANAGEMENT]: [EnumUserGroups.EXPAND, EnumUserGroups.MISSION_PLUS],
  [Routes.TAXONOMY_MANAGEMENT]: [EnumUserGroups.EXPAND, EnumUserGroups.MISSION_PLUS],
  [Routes.REPORT_MANAGEMENT]: [EnumUserGroups.EXPAND, EnumUserGroups.MISSION_PLUS],
  [Routes.COMPANY_MANAGEMENT]: [
    EnumUserGroups.EXPAND,
    EnumUserGroups.MISSION_PLUS,
    EnumUserGroups.EVS,
  ],
  [Routes.CHANGE_REQUEST_MANAGEMENT]: [
    EnumUserGroups.EXPAND,
    EnumUserGroups.MISSION_PLUS,
    EnumUserGroups.EVS,
  ],
  [Routes.INVESTOR_MANAGEMENT]: [
    EnumUserGroups.EXPAND,
    EnumUserGroups.MISSION_PLUS,
    EnumUserGroups.EVS,
  ],
  [Routes.MERGE_COMPANY]: [EnumUserGroups.EXPAND, EnumUserGroups.MISSION_PLUS, EnumUserGroups.EVS],
  [Routes.COMPANY_EDIT_SOURCE]: [
    EnumUserGroups.EXPAND,
    EnumUserGroups.MISSION_PLUS,
    EnumUserGroups.EVS,
  ],
  [Routes.SOURCE_MAPPING_MANAGEMENT]: [EnumUserGroups.EXPAND, EnumUserGroups.MISSION_PLUS],
  [Routes.ANALYSIS]: [EnumUserGroups.EXPAND, EnumUserGroups.MISSION_PLUS, EnumUserGroups.EVS],
}

export enum EnumAttachmentType {
  FCT_PROFILE = 'FCT profile',
  USE_CASE_PDF = 'Use case (PDF) - Live on portal',
  USE_CASE_PPT = 'Use case (PPT)',
  PRODUCT_PDF = 'Product (PDF) - Live on portal',
  FINANCIALS_PDF = 'Financials (PDF) - Live on portal',
  PITCH_DECK = 'Pitch deck',
  METHODOLOGY = 'Methodology',
  DEMO = 'Demo',
  OTHER = 'Other',
}

export enum TagTypes {
  FINTECHTYPE = 'FINTECHTYPE',
  TAG = 'TAG',
}

export enum EnumBoolean {
  TRUE = 1,
  FALSE = 0,
}

export enum EnumSourceNA {
  NA = 'NA',
}

export enum EnumInvestor {
  LEAD = 'Lead Investor',
  SUPPORT = 'Support Investor',
}

export enum ENumDataType {
  USE_CASE = 'use_cases',
  TECHNOLOGY = 'technologies',
  FUNDRAISING = 'fundraising',
}

export enum EnumSignUrlOperation {
  GET = 'getObject',
  PUT = 'putObject',
}

export enum EnumFileSize {
  PDF = 10485760,
  VIDEO = 52428800,
  IMG = 2097152,
}

export enum EPageKey {
  MAPPING_ZONE = 'mappingZone',
  FIND_FINTECHS = 'findFintechs',
  INCORRECT_MAPPING = 'incorrectMapping',
}

export enum ETaxonomyKeys {
  PRIMARY = 'primary',
  AUXILIARY = 'aux',
  GROUP = 'group',
}
