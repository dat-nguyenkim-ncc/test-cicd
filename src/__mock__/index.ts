import { TagGroupType } from './../types/index.d'
import {
  CompanyBusiness,
  CompanyDetail,
  CompanyFinancials,
  CompanyOverview,
  CompanyPeople,
  MappingSummary,
  TagData,
  TagDataParent,
} from '../types'
import { EnumAttachmentType, EnumCompanyTypeSector } from '../types/enums'

const getRandomId = (): string => {
  return (~~(Math.random() * 999999) + new Date().valueOf()).toString()
}

const companyTypes = Object.values(EnumCompanyTypeSector)
const suffix = ['Restaurant', 'Limited', 'UK', 'US', 'Bank']
export const mockCompanyId = getRandomId()
export const companyDetails = (): CompanyDetail => ({
  companyId: mockCompanyId,
  companyName: `Monzo ${suffix[~~(Math.random() * suffix.length)]}`,
  countryCode: 'GB',
  countryName: 'United Kingdom',
  url: 'https://monzo.com',
  companyTypes: [companyTypes[~~(Math.random() * companyTypes.length)]],
})

export const MockMonzoDetails = companyDetails()

export const mappingSummary = [
  {
    title: 'CATEGORIES',
    items: [
      { label: 'Fintech', parent: [] },
      { label: 'Retail Accounts' },
      { label: 'Accounts & Savings' },
      { label: 'Digital Banking' },
    ],
    type: EnumCompanyTypeSector.FIN,
  },
  {
    title: 'SECTORS',
    items: [{ label: 'No sector mapped', parent: [] }],
    type: EnumCompanyTypeSector.FIN,
  },
] as MappingSummary[]

export const companySources = [
  {
    company: MockMonzoDetails,
    source: {
      label: 'Source: XPTO',
      default: true,
    },
  },
  {
    company: companyDetails(),
    source: {
      label: 'Manual',
    },
  },
]

export const companyOverview = (): CompanyOverview => ({
  id: getRandomId(),
  source: 'CRB',
  sources: ['CRB', 'DR', 'MANUAL'],
  companyName: `Monzo ${suffix[~~(Math.random() * suffix.length)]}`,
  companyLocation: [
    {
      id: '123456',
      address: 'Broadwalk House<br>5 Appold Street<br>London<br>EC2A 2AG',
      postalCode: '94103',
      location: {
        region: 'EMEA',
        city: 'San Francisco',
        country: 'United States',
      },
      isHeadQuarter: true,
      selfDeclared: true,
      expandStatus: 'Following',
      source: 'bcg',
    },
  ],
  description: 'Monzo is a fintech challenger bank.',
  companyType: 'Startup',
  foundedYear: 2012,
  status: 'The company status',
  expandStatus: 'Following',
  expandStatusId: '1',
  closedDate: new Date(2012, 11, 1).toISOString(),
  ftse: 1000,
  logoUrl: 'https://monzo.com/logo.svg',
  contactEmail: 'monzo@monzo.com',
  phoneNumber: '0123456789',
  tags: null,
  otherNames: null,
  lastFundingType: null,
  logo_bucket_url: '',
  hashed_image: '',
  attachments: [
    {
      name: 'File nam',
      description: 'Broadwalk House<br>5 Appold Street<br>London<br>EC2A 2AG',
      type: EnumAttachmentType.FCT_PROFILE,
      url_attachment: '',
      expandStatus: 'Following',
      selfDeclared: true,
      date_created: '2021-03-30 10:43:33',
    },
  ],
  aliases: [
    {
      alias_id: '200354bcg1',
      company_alias: 'name-edit',
      expand_status_id: 1,
      selfDeclared: true,
      source: 'bcg',
    },
    {
      alias_id: '200354bcg2',
      company_alias: 'test1',
      expand_status_id: 1,
      selfDeclared: true,
      source: 'bcg',
    },
  ],
})

export const companyFinancials = (): CompanyFinancials => ({
  valuation: {
    value: 23123123,
    currency: 'USD',
  },
  fundingTotal: {
    value: 2312312312,
    currency: 'USD',
  },
  companyStage: 'Series E+',
  lastFundingDate: {
    value: 1231233,
    currency: 'USD',
  },
  equityFundingTotal: {
    value: 1231233,
    currency: 'USD',
  },
  leadInvestor: 'Mussum Cassilds-Cepacol',
  lastFundingAmount: {
    value: 123123123,
    currency: 'USD',
  },
  fundingRounds: [
    {
      id: '1',
      date: new Date(2012, 11, 1).toISOString(),
      quarter: 'Q2',
      comment: 'This is comment',
      roundTypes: ['Equity Financing, series E+'],
      expandRound1: 'Equity Financing',
      originExpandRound1: 'Equity Financing',
      valuation: {
        value: 123123123,
        currency: 'USD',
      },
      investment: {
        value: 123123123,
        currency: 'USD',
      },
      investors: {
        lead: 'Accel Partners, VC',
        other:
          'Goodwater Capital, VC Orange, Corporate Passion Capital, VC Reference Capital (formerly Genevest), VC Stripe, Tech Thrive Capital, VC Vanderbilt University, Other Y Combinator, Accelerator',
      },
    },
    {
      id: '2',
      date: new Date(2012, 11, 1).toISOString(),
      quarter: 'Q2',
      comment: 'This is comment',
      roundTypes: ['Equity Financing, series E+'],
      expandRound1: 'Equity Financing',
      originExpandRound1: 'Equity Financing',
      valuation: {
        value: 123123123,
        currency: 'USD',
      },
      investment: {
        value: 123123123,
        currency: 'USD',
      },
      investors: {
        lead: 'Accel Partners, VC',
      },
    },
    {
      id: '3',
      date: new Date(2012, 11, 1).toISOString(),
      quarter: 'Q2',
      comment: 'This is comment',
      roundTypes: ['Equity Financing, series E+'],
      expandRound1: 'Equity Financing',
      originExpandRound1: 'Equity Financing',
      valuation: {
        value: 123123123,
        currency: 'USD',
      },
      investment: {
        value: 123123123,
        currency: 'USD',
      },
      investors: {
        lead: 'Accel Partners, VC',
      },
    },
  ],
  isExternalViewDetail: true,
})

export const companyBusiness = (): CompanyBusiness => ({
  productsServices: ['Offers a current account and debit card '],
  businessRevenue: ['Service type: B2C'],
  targetClients: ['Retail customers'],
  partnerships: [
    '**Deliveroo:** Partnered with Deliveroo in Sep-2017 to allow users to split the bill on orders via its Monzo.me money request feature',
  ],
  vision: ['string'],
  keyMetrics: ['string'],
  differentiators: ['string'],
})

export const companyAcquisitions = () => []
export const companyIpos = () => []
export const companyPeople = (): CompanyPeople => ({ companyId: 145803 })

const tags = [
  'Capital Markets',
  'Corporate Banking',
  'Retail Banking',
  'SME Banking',
  'Support',
  'Technology',
  'Wealth Management',
]

const getRandomValueLabel = (index?: number) => {
  const tag = index ? tags[index] : tags[~~(Math.random() * tags.length)]
  return {
    label: tag,
    value: tag.toLowerCase().replace(' ', '_'),
  }
}

const getRandomBoolean = () => {
  return Math.random() * 10 > 5
}

export const TagMock = (
  index?: number,
  level: number = 0,
  parent: TagDataParent[] = []
): TagData => {
  const { value, label } = getRandomValueLabel(index)
  const id = getRandomId()
  const rowId = getRandomId()
  const isPriority = getRandomBoolean()
  return {
    id,
    rowId,
    value,
    label,
    isPriority,
    endpoint: level > 0 ? Math.random() > 0.5 : false,
    parent,
    description: 'Mock',
    children:
      level >= 5
        ? undefined
        : new Array(Math.max(3, ~~(Math.random() * tags.length)))
            .fill(null)
            .map((_a, index) =>
              TagMock(index, level + 1, [...parent, { id, rowId, label, value, isPriority }])
            ),
  }
}

export const ListRandomTags = (): TagData[] =>
  new Array(tags.length).fill(null).map((_a, index) => TagMock(index, 0))

const labels = [
  'Insurance Value Chain 1',
  'Insure Tech Tag',
  'Revenue model',
  'P&C Commercial Customers',
  'Product',
  'RegTech',
  'Regulation',
  'Report',
]

export const TagGroupMock = (): TagGroupType[] =>
  new Array(labels.length * 2).fill(null).map((_a, index) => {
    const id = getRandomId()
    const rowId = getRandomId()
    const label = labels[index % labels.length]
    const value = label.toLowerCase().split(' ').join('_')
    const isPriority = getRandomBoolean()
    return {
      id,
      rowId,
      label,
      value,
      isPriority,
      parent: [],
      children: new Array(Math.max(1, ~~(Math.random() * labels.length * 2)))
        .fill(null)
        .map((_b, index) => {
          const labelChild = labels[index % labels.length]
          return {
            label: labelChild,
            id: getRandomId(),
            rowId: getRandomId(),
            parent: [{ id, label, value }],
            isPriority,
            value: labelChild.toLowerCase().split(' ').join('_'),
          }
        }),
    }
  })
