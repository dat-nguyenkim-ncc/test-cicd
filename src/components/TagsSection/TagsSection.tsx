import React from 'react'
import { Box } from 'theme-ui'
import strings from '../../strings'
import { CompanyTypeSector, TagContainerParent, TagData, ViewInterface } from '../../types'
import { DimensionType, EnumDimensionType } from '../../types/enums'
import ButtonText from '../ButtonText'
import TagContainer from '../TagContainer'

export type TagsSectionProps = ViewInterface<{
  tags: TagData[]
  dimension: DimensionType
  type: Exclude<CompanyTypeSector, 'OUT'>
  selected?: TagData[]
  openedTags?: TagContainerParent
  onOpenTags(
    type: Exclude<CompanyTypeSector, 'OUT'>,
    dimension: DimensionType,
    parents: TagContainerParent
  ): void
  onSelectTags(type: Exclude<CompanyTypeSector, 'OUT'>, tags: TagData[]): void
  onClickUndo(type: Exclude<CompanyTypeSector, 'OUT'>, dimension: DimensionType): void
  unselectable?: TagData[]
  showUndo?: boolean
}>

export type TaxonomyHistory = ViewInterface<{
  selectedTags?: TagData[]
  openedTags?: TagContainerParent
}>

const TagsSection = ({
  sx,
  selected = [],
  openedTags = { tags: [], parent: [] },
  tags,
  dimension,
  type,
  onSelectTags,
  onOpenTags,
  onClickUndo,
  unselectable = [],
  showUndo = true,
}: TagsSectionProps) => {
  const onSelectTag = (tag: TagData) => {
    const newSelected =
      selected.filter(
        s =>
          s.id === tag.id &&
          s.parent.find(item => item.dimensionType === EnumDimensionType.SECTOR)?.id ===
            tag.parent.find(item => item.dimensionType === EnumDimensionType.SECTOR)?.id
      ).length > 0
        ? selected.filter(
            s =>
              !(
                s.id === tag.id &&
                s.parent.find(item => item.dimensionType === EnumDimensionType.SECTOR)?.id ===
                  tag.parent.find(item => item.dimensionType === EnumDimensionType.SECTOR)?.id
              )
          )
        : [...selected, { ...tag, link_id: `${Date.now()}` }]
    onSelectTags(type, newSelected)
  }

  const onOpenTag = (tag: TagData) => {
    let newTags = [...openedTags.tags]
    let newParent = [...openedTags.parent]

    // first level
    if (tags.map(t => t.id).includes(tag.id)) {
      newTags = []
      newParent = []
    }

    // second level
    const index = newParent
      .map(a => a?.id || undefined)
      .indexOf(tag.parent[tag.parent.length - 1]?.id)
    if (index > -1) {
      newTags.splice(index, newTags.length - index)
      newParent.splice(index, newParent.length - index)
    }

    newParent.push(tag.parent[tag.parent.length - 1])
    newTags = newTags.map(t => t.id).includes(tag.id)
      ? newTags.filter(t => t.id !== tag.id)
      : [...newTags, tag]

    onOpenTags(type, dimension, { parent: newParent, tags: newTags })
  }

  return (
    <Box sx={sx}>
      <TagContainer
        onSelect={onSelectTag}
        onOpen={onOpenTag}
        selected={selected}
        openedTags={openedTags?.tags || []}
        title={dimension}
        tags={tags}
        unselectable={unselectable}
      />
      {openedTags && openedTags.tags?.length > 0 && (
        <>
          {openedTags.tags.map((t, index) => (
            <TagContainer
              sx={{ mt: 4 }}
              key={index}
              onSelect={onSelectTag}
              onOpen={onOpenTag}
              selected={selected}
              openedTags={openedTags.tags || []}
              title={t.label}
              tags={t.children!}
              unselectable={unselectable}
            />
          ))}
        </>
      )}
      {showUndo && (
        <ButtonText
          onPress={() => {
            onClickUndo(type, dimension)
          }}
          label={strings.common.undoButton}
          sx={{ mt: 4 }}
        />
      )}
    </Box>
  )
}

export default TagsSection
