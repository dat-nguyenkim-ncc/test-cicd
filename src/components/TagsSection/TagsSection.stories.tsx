import React, { useState } from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'

import TagsSection, { TagsSectionProps } from './TagsSection'
import { ListRandomTags } from '../../__mock__'
import { Section } from '../primitives'
import {
  CompanyTypeSector,
  StateOpenedTags,
  StateSelectedTags,
  TagContainerParent,
  TagData,
} from '../../types'
import { DimensionType, EnumCompanyTypeSector, EnumDimensionType } from '../../types/enums'

export default {
  title: 'Taxonomy/TagsSection',
  component: TagsSection,
  parameters: {
    layout: 'fullscreen',
  },
} as Meta

const tags = ListRandomTags()
type Tabs = 'primary' | 'aux'

const Template: Story<TagsSectionProps> = args => {
  const [tabActive] = useState<Tabs>('primary')
  const [selectedTags, setSelectedTags] = useState<StateSelectedTags>()
  const [openedTags, setOpenedTags] = useState<StateOpenedTags>()
  const [selectedMap] = useState<Exclude<CompanyTypeSector, EnumCompanyTypeSector.OUT>>(
    EnumCompanyTypeSector.FIN
  )

  const onSelectTags = (type: string, tags: TagData[]) => {
    setSelectedTags({ ...selectedTags, [tabActive]: { [type]: tags } })
    console.log(JSON.stringify(tags))
  }

  const onOpenTags = (type: string, dimension: DimensionType, parents: TagContainerParent) => {
    setOpenedTags({ ...openedTags, [tabActive]: { [type]: parents } })
  }

  return (
    <Section sx={{ m: 4 }}>
      <TagsSection
        dimension={EnumDimensionType.SECTOR}
        type={EnumCompanyTypeSector.FIN}
        tags={tags}
        onSelectTags={onSelectTags}
        onOpenTags={onOpenTags}
        selected={
          selectedTags && selectedTags[tabActive] && selectedTags[tabActive]![selectedMap]
            ? selectedTags[tabActive]![selectedMap]
            : []
        }
        openedTags={
          openedTags &&
          openedTags[tabActive] &&
          openedTags[tabActive]![selectedMap]![EnumDimensionType.SECTOR]
            ? openedTags[tabActive]![selectedMap]![EnumDimensionType.SECTOR]
            : { parent: [], tags: [] }
        }
        onClickUndo={() => {}}
      />
    </Section>
  )
}

export const Default = Template.bind({})
