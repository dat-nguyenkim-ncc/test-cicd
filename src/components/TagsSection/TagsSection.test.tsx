import * as React from 'react'
import { shallow } from 'enzyme'
import TagsSection from './TagsSection'
import { ListRandomTags } from '../../__mock__'
import { TagData } from '../../types'
import { EnumCompanyTypeSector, EnumDimensionType } from '../../types/enums'

describe('TagsSection', () => {
  test('matches snapshot', () => {
    const wrapper = shallow(
      <TagsSection
        selected={[] as TagData[]}
        type={EnumCompanyTypeSector.FIN}
        openedTags={{ parent: [] as any, tags: [] as TagData[] }}
        dimension={EnumDimensionType.SECTOR}
        onOpenTags={() => {}}
        onSelectTags={() => {}}
        tags={ListRandomTags()}
        onClickUndo={() => {}}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
