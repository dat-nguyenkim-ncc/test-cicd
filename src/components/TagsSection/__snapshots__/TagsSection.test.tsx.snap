// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsSection matches snapshot 1`] = `ShallowWrapper {}`;
